src/readsums.c

```diff
@@ -56,18 +56,25 @@ static rs_result rs_loadsig_add_sum(rs_job_t *job, rs_strong_sum_t *strong)
     rs_signature_t      *sig = job->signature;
     rs_block_sig_t      *asignature;
 
-    sig->count++;
-    new_size = sig->count * sizeof(rs_block_sig_t);
+    if (sig->count <= job->stats.sig_blocks) {
```

Comment:
stats.sig_blocks is incremented for each signature block read.
sig->count is the estimated count of signatures (if set before based on the signature file size) or current count (if signature file size is not known). When sig->count <= job->stats.sig_blocks, we're out of memory (since at most sig->count blocks were allocated and we're trying to add a new block) so we need to reallocate, either because we miscalculated signature count or file size was not known in advance. This path is basically current path of librsync as if sig->count was never set in advance...